plugins/git4idea/src/git4idea/repo/GitRepositoryReader.java

```diff
@@ -492,10 +516,58 @@ private static boolean duplicateEntry(@NotNull Map<String, Hash> resolved, @NotN
   private static class HeadInfo {
     @Nullable private final String content;
     private final boolean isBranch;
+    @Nullable private final String symbolicRefName;
 
     HeadInfo(boolean branch, @Nullable String content) {
+      this(branch, content, null);
+    }
+
+    HeadInfo(boolean branch, @Nullable String content, @Nullable String symbolicRefName) {
       isBranch = branch;
       this.content = content;
+      this.symbolicRefName = symbolicRefName;
+    }
+
+    boolean isSymbolicRef() {
+      return symbolicRefName != null;
+    }
+  }
+
+  private static class RefInfo {
+    private final String refName;
+    private final Hash hash;
+    private final boolean isSymbolicRef;
```

Comment:
The boolean flag `isSymbolicRef` is not needed: it would be more reliable to declare a method 
```
boolean isSymbolicRef() { 
  return myRefencedRef != null; 
}
```